### ISSUE TITLE
Fiddling with auxiliary atomspace in peculiar way invalids a handle

### ISSUE THREAD

**Issue by ngeiswei** [opencog/atomspace#2382]
This one is wild. By accessing an atomspace via an atom in it, and re-accessing that handle makes it invalid. Here's the shortest way to reproduce it, I'll try to squeeze if I can but that's all I've got so far. Read carefully the comment in the end, as the bug can be circumvented by accessing the auxiliary atomspace directly instead of via an atom, really weird.

```scheme
(use-modules (opencog logger))
(cog-logger-set-timestamp! #f)
(cog-logger-set-stdout! #t)

;; Save initial atomspace
(define initial-as (cog-atomspace))

;; Define an auxiliary atomspace
(define auxiliary-as (cog-new-atomspace))

;; Switch to the auxiliary atomspace
(define pre-pre-as (cog-set-atomspace! auxiliary-as))

;; Inside this auxiliary atomspace, create concept A
(define auxiliary-A (Concept "A"))

;; Switch the atomspace where A is (which is auxiliary-as)
(define pre-as (cog-set-atomspace! (cog-as auxiliary-A)))

;; Verify that the current atomspace, auxiliary-as and the prevous one
;; are all the same
(cog-logger-info "[1] current as = ~a" (cog-atomspace))
(cog-logger-info "[1] auxiliary-as = ~a" auxiliary-as)
(cog-logger-info "[1] pre-as = ~a" pre-as)

;; Come back to the previous atomspace (which should still be auxiliary-as)
(cog-set-atomspace! pre-as)

;; Verify that the current atomspace and auxiliary-as are the same
(cog-logger-info "[2] current as = ~a" (cog-atomspace))
(cog-logger-info "[2] auxiliary-as = ~a" auxiliary-as)

;; Come back to the previous of the previous atomspace (which should be the initial atomspace)
(cog-set-atomspace! pre-pre-as)

;; Verify that the current atomspace and initial-as are the same
(cog-logger-info "[3] current as = ~a" (cog-atomspace))
(cog-logger-info "[3] initial-as = ~a" initial-as)

;; Switch again to the auxiliary atomspace
(define pre-as-again (cog-set-atomspace! auxiliary-as))

;; Create again concept A in it
(define A (Concept "A"))

;; Switch back to initial atomspace
(cog-set-atomspace! pre-as-again)

;; Now A is an invalid handle.
;;
;; Very strangly if the line
;;
;; (define pre-as (cog-set-atomspace! (cog-as auxiliary-A)))
;;
;; is replaced by
;;
;; (define pre-as (cog-set-atomspace! auxiliary-as))
;;
;; which should be equivalent, then A is no longer invalid.
```

**Comment by ngeiswei** [comment]
Also, I should point that sometimes it works! Especially right after purging guile cache. Otherwise it fails, that is A is invalid, the vast majority of the time.

I guess for now I'm gonna create a unit test. Perhaps I should create a guile and a C++ unit test to know whether the problem is in the scheme bindings or in the atomspace. I suspect the former, but have no certainty whatsoever.

**Comment by ngeiswei** [comment]
I don't seem to be able to make it fail in a unit test. Basically when run with `SchemeEval` is always passes, but if I load the same scheme file in guile it fails.

**Comment by ngeiswei** [comment]
The other odd and interesting thing I should mention is that if the last command

```scheme
;; Switch back to initial atomspace
(cog-set-atomspace! pre-as-again)
```

is disabled, then A is valid.

**Comment by ngeiswei** [comment]
OK, at the end of the script `SchemeSmob::release_as` is called and the auxiliary atomspace is deleted as a result. However, if

```scheme
(define pre-as (cog-set-atomspace! (cog-as auxiliary-A)))
```

is replaced by

```scheme
(define pre-as (cog-set-atomspace! auxiliary-as))
```

then the auxiliary atomspace is not released (which is the normal behavior)!

**Comment by ngeiswei** [comment]
I tried with guile-2.2.6, same problem.

I guess there would be a way to workaround that by hacking `SchemeSmob::as_ref_count`, not sure though, or maybe there's a way to prevent the gc from calling `SchemeSmob::release_as` in that instance.

**Comment by ngeiswei** [comment]
I tried with gc-8.0.4.tar.gz, same problem.

**Comment by ngeiswei** [comment]
OK, at least I managed to create a unit test that consistently fails, by triggering the gc and waiting, see `MultiAtomSpace::xtest_as_of_atom_scm()`.

**Comment by linas** [comment]
So, I'm cutting and pasting:
```
(use-modules (opencog logger))
(use-modules (opencog))
(cog-logger-set-timestamp! #f)
```
Crash!  Fixed in #2385
```
(cog-logger-set-stdout! #t)

;; Save initial atomspace
(define initial-as (cog-atomspace))

;; Define an auxiliary atomspace
(define auxiliary-as (cog-new-atomspace))

;; Switch to the auxiliary atomspace
(define pre-pre-as (cog-set-atomspace! auxiliary-as))
```
Of course ` pre-pre-as` should be the same as `initial-as` and we don't need the logger to see this: just printing will print the address of the atomspace in RAM (we could change this to alo print the UUID)
```
scheme@(guile-user)> initial-as
$3 = #<atomspace 0x5564368c2c30>
scheme@(guile-user)> auxiliary-as
$4 = #<atomspace 0x5564375dffd0>
scheme@(guile-user)> pre-pre-as
$5 = #<atomspace 0x5564368c2c30>
```
and pre-as should be sane as auxilliary-as:
```
scheme@(guile-user)> pre-as
$6 = #<atomspace 0x5564375dffd0>
```
This is misleading: the current as is the same as what you are setting it to:
```
;; Come back to the previous atomspace (which should still be auxiliary-as)
(cog-set-atomspace! pre-as)
```
which we can verify single by typing:
```
(cog-atomspace)
```
This this:
```
  ;; Now A is an invalid handle.
```
is the expected (defined) behavior: there is no "A" in the main AS; it exists in the aux AS only.

Next post explains how this works, under the covers.

**Comment by ngeiswei** [comment]
> > Now A is an invalid handle
> is the expected (defined) behavior: there is no "A" in the main AS; it exists in the aux AS only.

Are you sure it is the expected behavior? That seems very limiting, certainly the C++ API allows that. I mean I could rewrite my code to follow that behavior, but I'm not sure we want that.

**Comment by ngeiswei** [comment]
But if A being invalid is the expected behavior it should always be the case, regardless of whether the gc decided to release its atomspace. Also `auxiliary-as` is globally defined in the script, releasing it before the end is definitely wrong.

**Comment by linas** [comment]
Hang on ... I'm reviewing the code and it doesn't do what I thought it did.  So let me retract the "this s the expected (defined) behavior" statement for a moment ...

Yes, I can see how it would be nice to have guile symbols that reference atoms in other atomspaces, other than the current one.

**Comment by linas** [comment]
Yes, releasing `auxiliary-as` before the end is definitely wrong. There's some convoluted reference counting code for this, trying very hard to make sure it is not released.

**Comment by linas** [comment]
I cannot trigger the bug by cutting-n-pasting what you posted. Let me try `xtest_as_of_atom_scm`

**Comment by linas** [comment]
Interesting. If I cut-n-paste from `as-of-atom.scm` to the guile prompt, everything works. but if I run it by saying `(load "as-of-atom.scm")` then it fails. ...

This will require crawling over the atomspace reference-counting code, which is very definitely crazy-making. Do you want to do this, or should I?

**Comment by linas** [comment]
It's hard to see how this is not a guile bug.  My current working theory is that when `(load "as-of-atom.scm")` is done, that guile performs the loading in some kind of temporary environment. Once the load is done, the environment is discarded, and symbols that used to point at `auxiliary-as` are gc'ed.   Somehow, the gc does not notice that there are other symbols in the main environment that point to `auxiliary-as` and mistakenly tells opencog to release the AS. Which opencog happily obliges, as, so far as it can tell, the atomspace is needed unused.

**Comment by linas** [comment]
Yeah, seems to be submodule related ... I replaced `scm_eval_string` with `scm_eval_string_in_module((SCM)expr, scm_interaction_environment());` ... although it does not affect your unit tests, it does make `VectorAPIUTest` (and only that) fail, with crash because GC caused some atomspace to disappear.

**Comment by linas** [comment]
... or not. delete last comment, I screwed it up.

**Comment by linas** [comment]
Anyway, a work-around to this bug is this:
```--- a/opencog/guile/SchemeSmobAS.cc
+++ b/opencog/guile/SchemeSmobAS.cc
@@ -105,7 +105,13 @@ SCM SchemeSmob::ss_new_as (SCM s)
 
        // Only the internally-created atomspaces are trackable.
        std::lock_guard<std::mutex> lck(as_mtx);
-       deleteable_as[as] = 0;
+
+       // XXX FIXME. Setting `deleteable_as[as] = 1;` will cause a
+       // memory leak, allowing temporary atomspaces created in guile
+       // to pile up and never be garbage collected. However, this
+       // is a needed work-around for bug #2382.
+       // deleteable_as[as] = 0;
+       deleteable_as[as] = 1;
 
        // Guile might have discarded all references to the parent; but
        // we must not delete it, as long as a child is still referencing
```
It will prevent the accidental deletion of the auxiliary atomspace, and should make your tests pass.  At this time, this is a mostly-harmless work-around, unless you happen to be creating large numbers of temp atomspaces in scheme, in which case they will never be cleaned up, and remain clogging memory.

I'm exploring some more, but feel free to apply that patch, for now.
p.s. I am running `$ guile --version guile (GNU Guile) 2.9.2` and it's got this bug.

**Comment by linas** [comment]
Hmm. There are other problems.  If I just say `(cog-new-atomspace)` a zillion times, a zillion atomspaces will be created. They will never be gc'ed because there is nothing for guile to gc.

**Comment by linas** [comment]
This behaves unexpectedly:
```
(use-modules (opencog))
(define initial-as (cog-atomspace))
(define aux-as (cog-new-atomspace))  ;; call this aux-1
(set! aux-as (cog-new-atomspace))  ;; clobber aux-1, create aux-2
(cog-set-atomspace! aux-as)
(cog-set-atomspace! initial-as)
(set! aux-as (cog-new-atomspace))  ;; clobber aux-2, create aux-3
(set! aux-as 0)  ; clobber aux-3
(gc)
;; aux-1 and aux-3 are gc'ed but not aux-2
```
For some reason I do not understand, aux-2 is not gc'ed in the above.  I'm watching this with the patch below.
This seems to continue indefinitely:
```
(set! aux-as (cog-new-atomspace)) ;create aux-4
(cog-set-atomspace! aux-as) 
(cog-set-atomspace! initial-as)
(set! aux-as 0)   ; clobber aux-4
(gc)
```
So now, both aux-2 and aux-4 were not collected.  But the below it:
```
(set! aux-as (cog-new-atomspace))   ; create aux-5
(set! aux-as 0)
(gc)
```
and aux-5 is gc'ed.  So, somehow using `(cog-set-atomspace!` prevents gc from finding and collecting the dead as.  This is the *opposite* of the problem you reported (of live spaces getting gc'ed) ... but weirdly opposite.

But if I wrap the above in an env, then it is gc'ed ...:
```
(define (foo) (set! aux-as (cog-new-atomspace)) (cog-set-atomspace! aux-as) (cog-set-atomspace! initial-as) (set! aux-as 0))
(foo)
(gc)  ; the as that foo created does get gc'ed!
```
So, even though `foo` does exactly the same thing as the top-level evaluation environment, it behaves differently w.r.t gc.  This is confusing .... 

I'm watching the gc with print statements 
```
--- a/opencog/guile/SchemeSmobAS.cc
+++ b/opencog/guile/SchemeSmobAS.cc
@@ -67,8 +67,10 @@ void SchemeSmob::release_as (AtomSpace *as)
 {
        std::unique_lock<std::mutex> lck(as_mtx);
        auto has = deleteable_as.find(as);
+logger().info("duude release=%p has=%d", as, deleteable_as.end() == has);
        if (deleteable_as.end() == has) return;
 
+logger().info("duude release=%p cnt=%lu", as, deleteable_as[as]);
        if (0 == deleteable_as[as])
        {
                AtomSpace* env = as->get_environ();
@@ -107,6 +109,7 @@ SCM SchemeSmob::ss_new_as (SCM s)
        std::lock_guard<std::mutex> lck(as_mtx);
        deleteable_as[as] = 0;
 
+logger().info("duude create=%p with initial cnt=%lu", as, deleteable_as[as]);
        // Guile might have discarded all references to the parent; but
        // we must not delete it, as long as a child is still referencing
        // it.  The issue is that guile-gc does not know about this
@@ -305,6 +308,7 @@ SCM SchemeSmob::ss_get_as (void)
 void SchemeSmob::as_ref_count(SCM old_as, AtomSpace *nas)
 {
        AtomSpace* oas = ss_to_atomspace(old_as);
+logger().info("duude swap old=%p with new=%p", oas, nas);
        if (oas == nas) return;
 
        std::lock_guard<std::mutex> lck(as_mtx);
@@ -336,6 +340,10 @@ void SchemeSmob::as_ref_count(SCM old_as, AtomSpace *nas)
                        }
                }
        }
+if (deleteable_as.end() != deleteable_as.find(nas) and
+deleteable_as.end() != deleteable_as.find(oas))
+logger().info("duude done swap counts= old=%lu with new=%lu",
+deleteable_as[oas], deleteable_as[nas]);
 }
 
 /**
```

**Comment by linas** [comment]
OK, I think I can partly explain the above.  So, when you say
```
(define  bar "asdf")
```
guile create an "anonymous variable" that holds "asdf" and associates the symbol bar with that anonymous variable. If this anonymous variable is created in the top-level env, it lives forever, and is never gc'ed.  The contents of the variable might change, and the contents do get gc'ed, this is why aux-5 below is gc'ed.
```
(set! aux-as (cog-new-atomspace))   ; create aux-5
(set! aux-as 0)
(gc)
```
Anonymous variables in lower env's DO get garbage-collected. That's why aux-n below gets collected:
```
(define (foo) (set! aux-as (cog-new-atomspace)) ; create aux-n
     (cog-set-atomspace! aux-as) (cog-set-atomspace! initial-as)
     (set! aux-as 0))  ; clobber aux-n
(foo)
(gc) 
```
In the above, aux-n is gc'ed because the anoymous variable holding aux-n is gc'ed.

Finally we have aux-42 which is never collected:
```
(set! aux-as (cog-new-atomspace)) ; create aux-42
     (cog-set-atomspace! aux-as) (cog-set-atomspace! initial-as)
     (set! aux-as 0))  ; clobber aux-42
```
I'm not quite sure, but I think that whenever you say `(cog-set-atomspace! aux-as)` then a new anonymous variable is created in the eval environment, and this *new* variable is passed to `cog-set-atomspace!` and then this new variable is never gc'ed.  Since it is holding a pointer to aux-42, this results in aux-42 never being collected. Since its the top-level interaction environement, it is NEVER collected. See    https://stackoverflow.com/a/8482137/2565238 for explanation.

**Comment by linas** [comment]
So I think this explains why the copied-by-hand version of `as-of-atom.scm` works, while the `(load "as-of-atom")` fails.  Basically, there's a bunch of anonymous variables created during the load, and then they're cleaned up after the load completes.

And then, the final piece of the puzzle: After doing  `(load "as-of-atom") `, the main env now has symbols for `pre-as` and so on that are pointing at the gc'ed aux as. How is it that gc was so dumb as to collect the aux as when it was still obvious in-use?  Answer: because gc didn't know it was in use.  What gc is seeing is this:
```
SCM SchemeSmob::make_as(AtomSpace *as)
{
   SCM smob;
   SCM_NEWSMOB (smob, cog_misc_tag, as);
   SCM_SET_SMOB_FLAGS(smob, COG_AS);
   return smob;
}
```
which returns a smob with the address hashed into it, mangled/mashed so that gc doesn't know its a pointer to an atomspace.  So, in the `(load "as-of-atom") ` there is some SCM that holds `pre-as` and when load finishes, that SMOB 's gc'ed.  The fact that there is also *another* SMOB in the main env that also points to the same atomspae .. well, how can GC possibly know this? It cannot see the address.

So .. that explains *everything* I think.  I will now ponder a solution.

This is all very non-obvious, though.  I'm irked that the guile documentation doesn't really explain any of this.

**Comment by linas** [comment]
The fix would appear to be to do the counting in `make_as` instead of where it's currently counted.  This might be a bit conservative, but should work, I guess ...

**Comment by linas** [comment]
Heh. If you say
```
(cog-new-atomspace)
(cog-new-atomspace)
(cog-new-atomspace)
(gc)
```
The three atomspaces are NOT gc.ed

If you then say
```
(use-modules (ice-9 history))
(clear-value-history!)
(gc)
```
Then the three AtomSpaces ARE collected!  So now we know who was hanging on to those references. The history was!  See https://www.gnu.org/software/guile/manual/html_node/Value-History.html#Value-History
